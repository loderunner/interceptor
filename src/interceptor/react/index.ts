export { Provider as InterceptProvider, useIntercepts } from './intercept'
