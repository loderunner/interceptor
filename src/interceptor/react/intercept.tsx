import * as React from 'react'
import { createContext, useContext, useState } from 'react'

import * as intercept from '../intercept'

interface InterceptContextType {
  intercepts: intercept.Intercept[]
  addIntercept: (pattern: string) => void
  removeIntercept: (inter: intercept.Intercept) => void
}

const InterceptContext = createContext<InterceptContextType>({
  intercepts: [],
  addIntercept: () => {},
  removeIntercept: () => {},
})

interface Props {
  children: React.ReactNode
}

export const Provider = ({ children }: Props) => {
  const [intercepts, setIntercepts] = useState([...intercept.intercepts])

  const addIntercept = (pattern: string) => {
    intercept.addIntercept({ pattern, enabled: true })
    setIntercepts([...intercept.intercepts])
  }
  const removeIntercept = (inter: intercept.Intercept) => {
    intercept.removeIntercept(inter)
    setIntercepts([...intercept.intercepts])
  }

  return (
    <InterceptContext.Provider
      value={{ intercepts, addIntercept, removeIntercept }}
    >
      {children}
    </InterceptContext.Provider>
  )
}

export const useIntercepts = () => useContext(InterceptContext)
