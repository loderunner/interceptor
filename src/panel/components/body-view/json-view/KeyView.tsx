import * as React from 'react'
import { useCallback, useLayoutEffect, useRef, useState } from 'react'

import ModalInput from '../../ModalInput'

import type { SyntheticEvent } from 'react'

interface Props {
  name: string
  editable: boolean
  editingInitial?: boolean
  onChange?: (name: string) => void
  onCancel?: () => void
}

const KeyView = ({
  name,
  editable,
  editingInitial = false,
  onChange,
  onCancel,
}: Props) => {
  const [editing, setEditing] = useState(false)
  const preRef = useRef(null)

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useLayoutEffect(() => setEditing(editingInitial), [])

  const onDoubleClick = useCallback(
    (e: SyntheticEvent) => {
      if (!editable) {
        return
      }
      if (!preRef.current) {
        return
      }
      e.stopPropagation()
      setEditing(true)
    },
    [editable]
  )

  const onChangeInput = useCallback(
    (value) => {
      if (onChange !== undefined) {
        onChange(value)
      }
      setEditing(false)
    },
    [onChange]
  )

  const onCancelInput = useCallback(() => {
    if (onCancel !== undefined) {
      onCancel()
    }
    setEditing(false)
  }, [onCancel])

  return (
    <>
      <pre
        className="text-amber-700"
        onDoubleClick={onDoubleClick}
        ref={preRef}
      >
        {name}:
      </pre>{' '}
      {editing && preRef.current ? (
        <ModalInput
          element={preRef.current}
          value={name}
          onChange={onChangeInput}
          onCancel={onCancelInput}
        />
      ) : null}
    </>
  )
}

export default KeyView
