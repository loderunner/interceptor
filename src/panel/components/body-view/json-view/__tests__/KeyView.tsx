import { cleanup, render } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import * as React from 'react'
import '@testing-library/jest-dom'

import KeyView from '../KeyView'

describe('[JSON.KeyView]', () => {
  afterEach(() => {
    cleanup()
  })
  it('should match initial snapshot', () => {
    const { container } = render(<KeyView name="key" editable={false} />)
    expect(container).toMatchSnapshot()
  })

  it('should open text field on double-click', async () => {
    const user = userEvent.setup()
    const { getByText, queryByRole } = render(
      <KeyView name="key" editable={true} />
    )

    const pre = getByText('key:')

    await user.dblClick(pre)

    const textbox = queryByRole('textbox')
    expect(textbox).not.toBeNull()
    expect(textbox).toHaveValue('key')
  })

  it('should not open text field on double-click if not editable', async () => {
    const user = userEvent.setup()
    const { getByText, queryByRole } = render(
      <KeyView name="key" editable={false} />
    )

    const pre = getByText('key:')

    await user.dblClick(pre)

    const textbox = queryByRole('textbox')
    expect(textbox).toBeNull()
  })

  it('should not show text field on render', async () => {
    const { queryByRole } = render(<KeyView name="key" editable={true} />)

    const textbox = queryByRole('textbox')
    expect(textbox).toBeNull()
  })

  it('should show text field on render with editingInitial', async () => {
    const { queryByRole } = render(
      <KeyView name="key" editable={true} editingInitial={true} />
    )

    const textbox = queryByRole('textbox')
    expect(textbox).not.toBeNull()
  })

  it('should call onChange on close', async () => {
    const user = userEvent.setup()
    const onChange = jest.fn()
    const { getByText } = render(
      <KeyView name="key" editable={true} onChange={onChange} />
    )

    const pre = getByText('key:')

    await user.dblClick(pre)
    await user.keyboard('toto{Enter}')

    expect(onChange).toHaveBeenCalledWith('toto')
  })

  it('should call onCancel on Esc', async () => {
    const user = userEvent.setup()
    const onChange = jest.fn()
    const onCancel = jest.fn()
    const { getByText } = render(
      <KeyView
        name="key"
        editable={true}
        onChange={onChange}
        onCancel={onCancel}
      />
    )

    const pre = getByText('key:')

    await user.dblClick(pre)
    await user.keyboard('toto{Escape}')

    expect(onChange).not.toHaveBeenCalled()
  })
})
