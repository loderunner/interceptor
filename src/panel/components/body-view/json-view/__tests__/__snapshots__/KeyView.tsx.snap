// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[JSON.KeyView] should match initial snapshot 1`] = `
<div>
  <pre
    class="text-amber-700"
  >
    key
    :
  </pre>
   
</div>
`;
